src/include/btree_cmp_inline.h

```diff
@@ -14,12 +14,193 @@
+ *     user_item is lexicographically > tree_item. We use the names "user" and "tree" so it's clear
+ *     in the btree code which the application is looking at when we call its comparison function.
+ *     Some platforms like ARM offer dedicated instructions for reading 16 bytes at a time, allowing
+ *     for faster comparisons.
+ */
+static WT_INLINE int
+__lex_compare_gt_16(const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp)
+{
+    struct {
+        uint64_t a, b;
+    } tdata, udata;
+    uint64_t t64, u64;
+    const uint8_t *tendp, *treep, *uendp, *userp;
+    bool firsteq;
+
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+
+    /*
+     * Compare 16 bytes at a time until we find a difference or run out of 16 byte chunks to
+     * compare.
+     */
+    for (userp = ustartp, treep = tstartp; uendp - userp > WT_VECTOR_SIZE;
+         userp += WT_VECTOR_SIZE, treep += WT_VECTOR_SIZE) {
+        memcpy(&udata, userp, WT_VECTOR_SIZE);
+        memcpy(&tdata, treep, WT_VECTOR_SIZE);
+        if (udata.a != tdata.a || udata.b != tdata.b)
+            goto final128;
+    }
+
+    /*
+     * Rewind until there is exactly 16 bytes left. We know we started with at least 16, so we are
+     * still in bounds.
+     */
+    memcpy(&udata, uendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+    memcpy(&tdata, tendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+
+final128:
+    firsteq = udata.a == tdata.a;
+    u64 = firsteq ? udata.b : udata.a;
+    t64 = firsteq ? tdata.b : tdata.a;
+
+#ifndef WORDS_BIGENDIAN
+    u64 = __wt_bswap64(u64);
+    t64 = __wt_bswap64(t64);
+#endif
+
+    return (u64 < t64 ? -1 : u64 > t64 ? 1 : lencmp);
+}
+#endif
+
+/*
+ * __lex_compare_le_16 --
+ *     Lexicographic comparison routine for data less than or equal to 16 bytes. Returns: < 0 if
+ *     user_item is lexicographically < tree_item, = 0 if user_item is lexicographically =
+ *     tree_item, > 0 if user_item is lexicographically > tree_item. We use the names "user" and
+ *     "tree" so it's clear in the btree code which the application is looking at when we call its
+ *     comparison function. Some platforms like ARM offer dedicated instructions for reading 16
+ *     bytes at a time, allowing for faster comparisons.
+ */
+static WT_INLINE int
+__lex_compare_le_16(const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp)
+{
+    uint64_t ta, tb, ua, ub, u64, t64;
+    const uint8_t *tendp, *uendp;
+
+    if (len >= sizeof(uint64_t)) {
+        /*
+         * len >= 64 bits. len is implicitly less than or equal to 128bits since the function
+         * accepts 16 bytes or less.
+         */
+        uendp = ustartp + len;
+        tendp = tstartp + len;
+        memcpy(&ua, ustartp, sizeof(uint64_t));
+        memcpy(&ta, tstartp, sizeof(uint64_t));
+        memcpy(&ub, uendp - sizeof(uint64_t), sizeof(uint64_t));
+        memcpy(&tb, tendp - sizeof(uint64_t), sizeof(uint64_t));
```

Comment:
The algorithm here deserves some description. I think it breaks a number up to 128 bits into integer-comparable components. It uses memcpy to do that, since memcpy has optimizations that make it efficient for these small things, and avoids needing to carefully construct loops that need to check for complex termination conditions.

It would be best if there are comments explaining that clearly, rather than requiring future developers to reverse engineer it.